### PR TITLE
Some refactorings

### PR DESCRIPTION
**Pull request by soenkehahn**
Please, review #35 first as this branch builds on the other.

This PR consists of only loosely related commits, most of them only propositions. We should cherry-pick, what we agree on.

### CODE REVIEW

**Comment by soenkehahn** [comment]
`Servant.Common.Req.performRequest` is quite heavily refactored in this PR. But it's only internal so I just went ahead and implemented what I needed.

**Comment by alpmestan** [comment]
All the commits that were not in #35 look good to me. I could just merge this as is if you guys are ok with that?

File: servant/src/Servant/Utils/StaticFiles.hs [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+serveDirectory :: FilePath -> Server Raw
+serveDirectory documentRoot =
+  staticApp (defaultFileServerSettings (decodeString (documentRoot ++ "/")))
+
+serveDocumentation :: HasDocs api => Proxy api -> Server Raw
+serveDocumentation proxy _request respond =
+  respond $ responseLBS ok200 [] $ cs $ toHtml $ markdown $ docs proxy
+
+toHtml :: String -> String
+toHtml markdown =
```
Comment (jkarni) [inline servant/src/Servant/Utils/StaticFiles.hs:26]: not a big deal, but this shadows an existing binding (I'm in favor of not having anything come out of `-Wall`)

File: servant/src/Servant/Docs.hs [hunk 2]
```diff
@@ -349,69 +358,80 @@ class ToCapture c where
...
-
-  where printEndpoint endpoint action = do
-          putStrLn $ str
-          putStrLn $ replicate len '-'
-          putStrLn ""
-          capturesStr (action ^. captures)
-          paramsStr (action ^. params)
-          rqbodyStr (action ^. rqbody)
-          responseStr (action ^. response) 
+printMarkdown = print . markdown
```
Comment (jkarni) [inline servant/src/Servant/Docs.hs]: shouldn't this be `putStr` or `putStrLn`? It'd look better

Comment (alpmestan) [inline servant/src/Servant/Docs.hs]: Yup, it was but Soenke switched to `print`.

File: servant/src/Servant/Utils/StaticFiles.hs (hunk 1 above)
Comment (soenkehahn) [inline servant/src/Servant/Utils/StaticFiles.hs:26]: I'm in favor of having _some_ warnings settings and stick to the policy that we don't want warnings. (In my experience we shouldn't have `-Werror` as sometimes we want warnings and still commit them, e.g. for deprecations.) For my own projects I'm mostly using `-Wall -fno-warn-name-shadowing`, but I'm perfectly fine with `-Wall` for servant. (There's lots of name shadowing going on, that we'd have to fix.)

File: servant/src/Servant/Docs.hs (hunk 2 above)
Comment (soenkehahn) [inline servant/src/Servant/Docs.hs]: Yes, of course. I was confused by the name `printMarkdown`.

Comment (soenkehahn) [inline servant/src/Servant/Docs.hs]: I was thinking of removing `printMarkdown` altogether. It's easy to just write `putStrLn . markdown` yourself.

(Well, I'm one to talk given that I messed that up. :))

What do you think?

Comment (jkarni) [inline servant/src/Servant/Docs.hs]: :+1:

File: servant/src/Servant/Utils/StaticFiles.hs (hunk 1 above)
Comment (jkarni) [inline servant/src/Servant/Utils/StaticFiles.hs:26]: Sorry, I should have been clearer. I'm in favor of no warnings showing up when we compile. `fno-warn-name-shadowing` is fine by me (at the file or cabal level).

File: servant/src/Servant/Docs.hs (hunk 2 above)
Comment (alpmestan) [inline servant/src/Servant/Docs.hs]: Yup, sounds fine :)

**Comment by soenkehahn** [comment]
Good to merge, pending travis.

File: servant/src/Servant/Utils/StaticFiles.hs (hunk 1 above)
Comment (soenkehahn) [inline servant/src/Servant/Utils/StaticFiles.hs:26]: I agree. Not overly passionate about `-fno-warn-name-shadowing`.

Comment (soenkehahn) [inline servant/src/Servant/Utils/StaticFiles.hs:26]: See also #38.

**Pull request merged**